File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
package com.parsely.parselyandroid;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;
import java.util.TimeZone;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Engagement manager for article and video engagement.
 * <p>
 * Implemented to handle its own queuing of future executions to accomplish
 * two things:
 * <p>
 * 1. Flushing any engaged time before canceling.
 * 2. Progressive backoff for long engagements to save data.
 */
class EngagementManager {

    private final ParselyTracker parselyTracker;
    public Map<String, Object> baseEvent;
    private boolean started;
    private final Timer parentTimer;
    private TimerTask waitingTimerTask;
    private long latestDelayMillis, totalTime;
    private Calendar startTime;



    public EngagementManager(ParselyTracker parselyTracker, Timer parentTimer, long intervalMillis, Map<String, Object> baseEvent) {
        this.parselyTracker = parselyTracker;
        this.baseEvent = baseEvent;
        this.parentTimer = parentTimer;
        latestDelayMillis = intervalMillis;
        totalTime = 0;
        startTime = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
    }

    public boolean isRunning() {
        return started;
    }

    public void start() {
        scheduleNextExecution(latestDelayMillis);
        started = true;
        startTime = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
    }

    public void stop() {
        waitingTimerTask.cancel();
        started = false;
    }

    public boolean isSameVideo(String url, String urlRef, ParselyVideoMetadata metadata) {
        Map<String, Object> baseMetadata = (Map<String, Object>) baseEvent.get("metadata");
        return (baseEvent.get("url").equals(url) &&
                baseEvent.get("urlref").equals(urlRef) &&
                baseMetadata.get("link").equals(metadata.link) &&
                (int) (baseMetadata.get("duration")) == metadata.durationSeconds);
    }

    private void scheduleNextExecution(long delay) {
        TimerTask task = new TimerTask() {
            public void run() {
                doEnqueue(scheduledExecutionTime());
                updateLatestInterval();
                scheduleNextExecution(latestDelayMillis);
            }

            public boolean cancel() {
                boolean output = super.cancel();
                // Only enqueue when we actually canceled something. If output is false then
                // this has already been canceled.
                if (output) {
                    doEnqueue(scheduledExecutionTime());
                }
                return output;
            }
        };
        latestDelayMillis = delay;
        parentTimer.schedule(task, delay);
        waitingTimerTask = task;
    }

    private void doEnqueue(long scheduledExecutionTime) {
        // Create a copy of the base event to enqueue
        Map<String, Object> event = new HashMap<>(baseEvent);
        ParselyTracker.PLog(String.format("Enqueuing %s event.", event.get("action")));

        // Update `ts` for the event since it's happening right now.
        Calendar now = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        @SuppressWarnings("unchecked")
        Map<String, Object> baseEventData = (Map<String, Object>) event.get("data");
        assert baseEventData != null;
        Map<String, Object> data = new HashMap<>((Map<String, Object>) baseEventData);
        data.put("ts", now.getTimeInMillis());
        event.put("data", data);

        // Adjust inc by execution time in case we're late or early.
        long executionDiff = (System.currentTimeMillis() - scheduledExecutionTime);
        long inc = (latestDelayMillis + executionDiff);
        totalTime += inc;
        event.put("inc", inc / 1000);
        event.put("tt", totalTime);

        parselyTracker.enqueueEvent(event);
    }


    public double getIntervalMillis() {
        return latestDelayMillis;
    }
}

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.kt
package com.parsely.parselyandroid

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Engagement manager for article and video engagement.
 *
 *
 * Implemented to handle its own queuing of future executions to accomplish
 * two things:
 *
 *
 * 1. Flushing any engaged time before canceling.
 * 2. Progressive backoff for long engagements to save data.
 */
internal class EngagementManager(
    private val eventQueuer: EventQueuer,
    private var latestDelayMillis: Long,
    private val baseEvent: Map<String, Any>,
    private val intervalCalculator: HeartbeatIntervalCalculator,
    private val coroutineScope: CoroutineScope,
    private val clock: Clock,
) {
    private var job: Job? = null
    private var totalTime: Long = 0
    private var nextScheduledExecution: Long = 0

    val isRunning: Boolean
        get() = job?.isActive ?: false

    fun start() {
        val startTime = clock.now
        job = coroutineScope.launch {
            while (isActive) {
                latestDelayMillis = intervalCalculator.calculate(startTime)
                nextScheduledExecution = clock.now.inWholeMilliseconds + latestDelayMillis
                delay(latestDelayMillis)
                doEnqueue(clock.now.inWholeMilliseconds)
            }
        }
    }

    fun stop() {
        job?.let {
            it.cancel()
            doEnqueue(nextScheduledExecution)
        }
    }

    fun isSameVideo(url: String, urlRef: String, metadata: ParselyVideoMetadata): Boolean {
        val baseMetadata = baseEvent["metadata"] as Map<String, Any>?
        return baseEvent["url"] == url && baseEvent["urlref"] == urlRef && baseMetadata!!["link"] == metadata.link && baseMetadata["duration"] as Int == metadata.durationSeconds
    }

    private fun doEnqueue(scheduledExecutionTime: Long) {
        // Create a copy of the base event to enqueue
        val event: MutableMap<String, Any> = HashMap(
            baseEvent
        )
        Log.d("Enqueuing ${event["action"]} event.")

        // Update `ts` for the event since it's happening right now.
        val baseEventData = (event["data"] as Map<String, Any>?)!!
        val data: MutableMap<String, Any> = HashMap(baseEventData)
        data["ts"] = clock.now.inWholeMilliseconds
        event["data"] = data

        // Adjust inc by execution time in case we're late or early.
        val executionDiff = clock.now.inWholeMilliseconds - scheduledExecutionTime
        val inc = latestDelayMillis + executionDiff
        totalTime += inc
        event["inc"] = inc / 1000
        event["tt"] = totalTime
        eventQueuer.enqueueEvent(event)
    }

    val intervalMillis: Double
        get() = latestDelayMillis.toDouble()
}
